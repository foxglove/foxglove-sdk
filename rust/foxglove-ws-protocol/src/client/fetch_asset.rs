use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// Fetch asset message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#fetch-asset>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "fetchAsset", rename_all = "camelCase")]
pub struct FetchAsset {
    /// Asset URI.
    pub uri: String,
    /// Request ID.
    pub request_id: u32,
}

impl JsonMessage for FetchAsset {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;

    use super::*;

    fn message() -> FetchAsset {
        FetchAsset {
            uri: "package://foxglove/example.urdf".to_string(),
            request_id: 42,
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::FetchAsset(orig));
    }
}
