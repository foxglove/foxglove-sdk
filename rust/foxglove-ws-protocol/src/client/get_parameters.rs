use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// Get parameters message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#get-parameters>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "getParameters", rename_all = "camelCase")]
pub struct GetParameters {
    /// Parameter names.
    pub parameter_names: Vec<String>,
    /// Request ID.
    pub id: Option<String>,
}

impl JsonMessage for GetParameters {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;

    use super::*;

    fn message() -> GetParameters {
        GetParameters {
            parameter_names: vec!["param1".to_string(), "param2".to_string()],
            id: Some("request-123".to_string()),
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::GetParameters(orig));
    }
}
