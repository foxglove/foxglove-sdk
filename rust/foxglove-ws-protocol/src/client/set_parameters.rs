use serde::{Deserialize, Serialize};

use crate::parameter::Parameter;
use crate::JsonMessage;

/// Set parameters message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#set-parameters>
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "setParameters", rename_all = "camelCase")]
pub struct SetParameters {
    /// Parameters.
    pub parameters: Vec<Parameter>,
    /// Request ID.
    pub id: Option<String>,
}

impl JsonMessage for SetParameters {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;
    use crate::parameter::Parameter;

    use super::*;

    fn message() -> SetParameters {
        SetParameters {
            parameters: vec![
                Parameter::empty("empty"),
                Parameter::float64("f64", 1.23),
                Parameter::float64_array("f64[]", vec![1.23, 4.56]),
                Parameter::byte_array("byte[]", [0x10, 0x20, 0x30]),
                Parameter::bool("bool", true),
            ],
            id: Some("request-123".to_string()),
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::SetParameters(orig));
    }
}
