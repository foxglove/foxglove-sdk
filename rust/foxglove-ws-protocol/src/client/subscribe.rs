//! Subscribe message types.

use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// Subscribe message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#subscribe>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "subscribe", rename_all = "camelCase")]
pub struct Subscribe {
    /// Subscriptions.
    pub subscriptions: Vec<Subscription>,
}

/// A subscription for a [`Subscribe`] message.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Subscription {
    /// Subscription ID.
    pub id: u32,
    /// Channel ID.
    pub channel_id: u64,
}

impl JsonMessage for Subscribe {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;

    use super::*;

    fn message() -> Subscribe {
        Subscribe {
            subscriptions: vec![
                Subscription {
                    id: 1,
                    channel_id: 10,
                },
                Subscription {
                    id: 2,
                    channel_id: 20,
                },
            ],
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::Subscribe(orig));
    }
}
