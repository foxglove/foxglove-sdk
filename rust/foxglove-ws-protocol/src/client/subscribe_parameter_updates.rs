use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// Subscribe parameter updates message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#subscribe-parameter-update>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(
    tag = "op",
    rename = "subscribeParameterUpdates",
    rename_all = "camelCase"
)]
pub struct SubscribeParameterUpdates {
    /// Parameter names.
    pub parameter_names: Vec<String>,
}

impl SubscribeParameterUpdates {
    /// Creates a new SubscribeParameterUpdates from an iterator of strings.
    pub fn new(names: impl IntoIterator<Item = impl Into<String>>) -> Self {
        Self {
            parameter_names: names.into_iter().map(Into::into).collect(),
        }
    }
}

impl JsonMessage for SubscribeParameterUpdates {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;

    use super::*;

    fn message() -> SubscribeParameterUpdates {
        SubscribeParameterUpdates {
            parameter_names: vec!["param1".to_string(), "param2".to_string()],
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::SubscribeParameterUpdates(orig));
    }
}
