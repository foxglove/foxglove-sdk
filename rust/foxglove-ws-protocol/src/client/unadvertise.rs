use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// Client unadvertise message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#client-unadvertise>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "unadvertise", rename_all = "camelCase")]
pub struct Unadvertise {
    /// Channel IDs.
    pub channel_ids: Vec<u32>,
}

impl JsonMessage for Unadvertise {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;

    use super::*;

    fn message() -> Unadvertise {
        Unadvertise {
            channel_ids: vec![1, 2, 3],
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::Unadvertise(orig));
    }
}
