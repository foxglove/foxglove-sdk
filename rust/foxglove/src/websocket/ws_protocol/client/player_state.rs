use bytes::{Buf, BufMut};

use crate::websocket::ws_protocol::{BinaryMessage, ParseError};

use super::BinaryOpcode;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum PlaybackState {
    Playing = 0,
    Paused = 1,
    Buffering = 2,
    Ended = 3,
}

impl TryFrom<u8> for PlaybackState {
    type Error = u8;

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(Self::Playing),
            1 => Ok(Self::Paused),
            2 => Ok(Self::Buffering),
            3 => Ok(Self::Ended),
            _ => Err(value),
        }
    }
}

/// Player state message.
#[derive(Debug, Clone, PartialEq)]
pub struct PlayerState {
    /// Playback state
    pub playback_state: PlaybackState,
    /// Playback speed
    pub playback_speed: f32,
    /// Seek playback time in nanoseconds (only set if a seek has been performed)
    pub seek_time: Option<u64>,
}

impl<'a> BinaryMessage<'a> for PlayerState {
    fn parse_binary(mut data: &'a [u8]) -> Result<Self, ParseError> {
        // Message size: playback_state (1 byte) + playback_speed (4 bytes) + had_seek (1 byte) + seek_time (8 bytes)
        if data.len() < 1 + 4 + 1 + 8 {
            return Err(ParseError::BufferTooShort);
        }

        let state_byte = data.get_u8();
        let playback_state = PlaybackState::try_from(state_byte)
            .map_err(|_| ParseError::InvalidPlaybackState(state_byte))?;

        let playback_speed = data.get_f32_le();
        let had_seek = data.get_u8() != 0;
        let seek_time = if had_seek && data.len() >= 8 {
            Some(data.get_u64_le())
        } else {
            None
        };

        Ok(Self {
            playback_state,
            playback_speed,
            seek_time,
        })
    }

    fn to_bytes(&self) -> Vec<u8> {
        // Message size: opcode (1) + playback_state (1) + playback_speed (4) + had_seek (1) + seek_time (8)
        let mut buf = Vec::with_capacity(1 + 4 + 1 + 8);

        buf.put_u8(BinaryOpcode::PlayerState as u8);
        buf.put_u8(self.playback_state as u8);
        buf.put_f32_le(self.playback_speed);
        buf.put_u8(if self.seek_time.is_some() { 1 } else { 0 });
        // To keep the message size constant, write the seek time even if it's None.
        buf.put_u64_le(self.seek_time.unwrap_or(0));
        buf
    }
}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::client::ClientMessage;

    use super::*;

    #[test]
    fn test_encode_playing() {
        let message = PlayerState {
            playback_state: PlaybackState::Playing,
            playback_speed: 1.0,
            seek_time: None,
        };
        insta::assert_snapshot!(format!("{:#04x?}", message.to_bytes()));
    }

    #[test]
    fn test_encode_paused() {
        let message = PlayerState {
            playback_state: PlaybackState::Paused,
            playback_speed: 1.0,
            seek_time: None,
        };
        insta::assert_snapshot!(format!("{:#04x?}", message.to_bytes()));
    }

    #[test]
    fn test_encode_buffering() {
        let message = PlayerState {
            playback_state: PlaybackState::Buffering,
            playback_speed: 1.0,
            seek_time: None,
        };
        insta::assert_snapshot!(format!("{:#04x?}", message.to_bytes()));
    }

    #[test]
    fn test_encode_ended() {
        let message = PlayerState {
            playback_state: PlaybackState::Ended,
            playback_speed: 1.0,
            seek_time: None,
        };
        insta::assert_snapshot!(format!("{:#04x?}", message.to_bytes()));
    }

    #[test]
    fn test_roundtrip_playing_with_seek_time() {
        let orig = PlayerState {
            playback_state: PlaybackState::Playing,
            playback_speed: 1.0,
            seek_time: Some(100_500_000_000),
        };
        let buf = orig.to_bytes();
        let msg = ClientMessage::parse_binary(&buf).unwrap();
        assert_eq!(msg, ClientMessage::PlayerState(orig));
    }

    #[test]
    fn test_roundtrip_playing_without_seek_time() {
        let orig = PlayerState {
            playback_state: PlaybackState::Playing,
            playback_speed: 1.0,
            seek_time: None,
        };
        let buf = orig.to_bytes();
        let msg = ClientMessage::parse_binary(&buf).unwrap();
        assert_eq!(msg, ClientMessage::PlayerState(orig));
    }

    #[test]
    fn test_parse_binary_with_seek_time() {
        // Manually construct binary data: opcode + state + speed + had_seek + seek_time
        let mut data = Vec::new();
        data.put_u8(BinaryOpcode::PlayerState as u8); // opcode
        data.put_u8(PlaybackState::Playing as u8); // state
        data.put_f32_le(1.5); // speed
        data.put_u8(1); // had_seek = true
        data.put_u64_le(100_500_000_000); // seek_time

        let msg = ClientMessage::parse_binary(&data).unwrap();
        match msg {
            ClientMessage::PlayerState(state) => {
                assert_eq!(state.playback_state, PlaybackState::Playing);
                assert_eq!(state.playback_speed, 1.5);
                assert_eq!(state.seek_time, Some(100_500_000_000));
            }
            _ => panic!("Expected PlayerState message"),
        }
    }

    #[test]
    fn test_parse_binary_without_seek_time() {
        // Manually construct binary data with had_seek = false (seek_time bytes still present but zeroed)
        let mut data = Vec::new();
        data.put_u8(BinaryOpcode::PlayerState as u8); // opcode
        data.put_u8(PlaybackState::Buffering as u8); // state
        data.put_f32_le(2.0); // speed
        data.put_u8(0); // had_seek = false
        data.put_u64_le(0); // seek_time (zeroed out, ignored since had_seek = false)

        let msg = ClientMessage::parse_binary(&data).unwrap();
        match msg {
            ClientMessage::PlayerState(state) => {
                assert_eq!(state.playback_state, PlaybackState::Buffering);
                assert_eq!(state.playback_speed, 2.0);
                assert_eq!(state.seek_time, None);
            }
            _ => panic!("Expected PlayerState message"),
        }
    }
}
