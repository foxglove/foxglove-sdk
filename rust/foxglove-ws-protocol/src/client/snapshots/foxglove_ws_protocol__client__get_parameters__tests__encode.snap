---
source: src/client/get_parameters.rs
expression: message()
---
{
  "op": "getParameters",
  "parameterNames": [
    "param1",
    "param2"
  ],
  "id": "request-123"
}
